### PR TITLE
Add HIP for validator challenges

### PR DESCRIPTION
**Pull request by Vagabond**
[Rendered](https://github.com/helium/HIP/blob/adt/validator-challenges/xxxx-validator-challenges.md)

### CODE REVIEW

File: xxxx-validator-challenges.md [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+refer to the Detailed Explanation section of the [H3Dex Targeting HIP](link when
+merged) which describes how a challenge is constructed and submitted to the
+chain.
+
+Once the PoC request has appeared on chain and the Challengee has been selected
+the Hotspot attempts to connect to it over p2p (failure point #1). If successful
+it sends the challengee the data to transmit. The Challengee then consults its
+ledger to see if the PoC is active (failure point #2 if the Challengee is not
+synced). The Challengee then transmits the packet and sends a receipt to the
+Challengeer over the same connection the Challenger dialed on (this avoids
```
Comment (andymck) [inline xxxx-validator-challenges.md:65]:
```suggestion
Challenger over the same connection the Challenger dialled on (this avoids
```

File: xxxx-validator-challenges.md [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+light gateways will be streamed chain related data, events and notifications.
+
+Light gateways can also connect to additional validators as required.  Such
+connections will be ephemeral in nature and typically for unary requests.
+
+All message types sent to a light gateway from a validator will be wrapped in a
+top level msg of the type 'gateway_resp_v1' ( see proto link below ).  This msg
+type includes a signed payload of the msg and in addition the block height at
+which the msg contents were valid.  Together these two fields serve as an
+attestation on behalf of the sending validator.
```
Comment (andymck) [inline xxxx-validator-challenges.md:102]:
```suggestion
All message types sent to a light gateway from a validator will be wrapped in a
top level msg of the type 'gateway_resp_v1'.  This msg
type includes a number of fields which together serve as an attestation on behalf of the sending validator.  The attestation data includes:
- A signed payload of the msg
- Block age
- Block Height
- Block Time ```

File: xxxx-validator-challenges.md [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+generated.  The region, target and challenge will be persisted to disk.
+
+### Validators & Light Gateway Communication
+
+All validators, whether they are in consensus or not, as part of handling each
+block will pull the public key hashes from the block metadata and together with
+the block hash generate the same entropy as the CG members to identify the
+target region.  The validators will then pull a list of all gateways within that
+region and send each gateway in that region connected to them a notification msg
+of type 'poc_challenge_notification_resp_v1' informing them of a challenge
```
Comment (andymck) [inline xxxx-validator-challenges.md:132]:
```suggestion
 informing them of a challenge
```

File: xxxx-validator-challenges.md [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+within their region. The payload  provides the necessary routing data ( pubkey
```
Comment (andymck) [inline xxxx-validator-challenges.md:133]:
```suggestion
within their region. The notification  provides the necessary routing data ( pubkey
```

File: xxxx-validator-challenges.md [hunk 5]
```diff
@@ -0,0 +1,304 @@
...
+and IP) to enable the light gateway to connect to the challenging validator and
+in addition includes the onion key hash of the challenge.
+
+All gateways upon receipt of a challenge notification will respond with a unary
+GRPC request of type 'gateway_poc_check_challenge_target_req_v1' to the
+challenging CG member to check if they are the target.
```
Comment (andymck) [inline xxxx-validator-challenges.md:139]:
```suggestion
All gateways upon receipt of a challenge notification will send a request over GRPC to the
challenging Validator to check if they are the target.
```

File: xxxx-validator-challenges.md (hunk 1 above)
Comment (abhay) [reply xxxx-validator-challenges.md:65]: resolved in #361

File: xxxx-validator-challenges.md (hunk 2 above)
Comment (abhay) [reply xxxx-validator-challenges.md:102]: Addressed in #361

File: xxxx-validator-challenges.md (hunk 4 above)
Comment (abhay) [reply xxxx-validator-challenges.md:133]: resolved in #361

File: xxxx-validator-challenges.md (hunk 5 above)
Comment (abhay) [reply xxxx-validator-challenges.md:139]: Resolved in #361

File: xxxx-validator-challenges.md (hunk 3 above)
Comment (abhay) [reply xxxx-validator-challenges.md:132]: Resolved in #361

File: xxxx-validator-challenges.md [hunk 6]
```diff
@@ -0,0 +1,319 @@
...
+harder to source capable hardware for building Hotspots that can meet these
+requirements.
+
+To address these issues the core developers have been working on design and
+implementation of an alternative PoC Challenge mechanism we call Validator
+Challenges. In brief, Validator Challenges move the role of generating
+Challenges to the Consensus Group. This not only allows us to free Hotspots from
+the burden of following the blockchain but it also moves the entities initiating
+Challenges to machines with much more stable and predictable networking which
+reduces the likihood of connectivity failures. Hotspots can become clients of
```
Comment (cokes518) [inline xxxx-validator-challenges.md:43]:
```suggestion
reduces the likelihood of connectivity failures. Hotspots can become clients of
```

File: xxxx-validator-challenges.md [hunk 7]
```diff
@@ -0,0 +1,319 @@
...
+how a challenge is constructed and submitted to the blockchain.
+
+1. Once the PoC Request has appeared on the blockchain and the Challengee has
+   been selected, the Hotspot attempts to connect to it over p2p (potential
+   failure point). If successful it sends the challengee the data to transmit.
+2. The Challengee then consults its local blockchain ledger to see if the PoC is
+   active. This is another potential failure point if the Challengee is not
+   synced.
+3. The Challengee then transmits the packet and sends a receipt to the
+   Challenger over the same connection from the Challenger. Tthis avoids another
```
Comment (cokes518) [inline xxxx-validator-challenges.md:68]:
```suggestion
   Challenger over the same connection from the Challenger. This is another
```

File: xxxx-validator-challenges.md [hunk 8]
```diff
@@ -0,0 +1,319 @@
...
+   potential p2p failure.
+4. Any Hotspots observing the challenge packet then consults their local
+   blockchain ledger to see if they can resolve the challenge to a Challenger.
+   This is another potential failure point if the witness is not in sync with
+   the blockchain.
+5. If the challenge can be resolved to a Challenger then the Witness attempts to
+   dial that Challenger over p2p, another potential failure point.
+6. Once the Challenger has decided enough blocks have passed it submits the
+   `poc_receipt` transaction to the blockchain to report on the information
+   gathered (if any. The final, potential failure point is if the Challenger is
```
Comment (cokes518) [inline xxxx-validator-challenges.md:78]:
```suggestion
   gathered (if any). The final, potential failure point is if the Challenger is
```

File: xxxx-validator-challenges.md [hunk 9]
```diff
@@ -0,0 +1,319 @@
...
+
+An important aspect of this system is that the actual challenge entropy is kept
+secret until the receipt transaction is published. This requires Hotspots to
+observe the actual data from the Challenger to correctly respond to the
+challenge and prevents poisoning attacks. However Hotspots can verify the
+challenge information against the blockchain once they see it.
+
+## Light Hotspots & GRPC
+
+Light Hotspots will maintain a durable GRPC connection to one (or many)
```
Comment (cokes518) [inline xxxx-validator-challenges.md:96]:
```suggestion
Light Hotspots will maintain a durable GRPC (Google Remote Procedure Call) connection to one (or many)
```

File: xxxx-validator-challenges.md [hunk 10]
```diff
@@ -0,0 +1,319 @@
...
+Validators.  The target Validator(s) could be random or specified.  Via the
+first durable connection, Light Hotspots will be streamed chain related data,
+events and notifications. Additional connections can be considered ephermeral
+and used for lookup requests if neded.
```
Comment (cokes518) [inline xxxx-validator-challenges.md:100]:
```suggestion
and used for lookup requests if needed.
```

File: xxxx-validator-challenges.md [hunk 11]
```diff
@@ -0,0 +1,319 @@
...
+of an issue, but with over 500,000 Hotspots on the network now and larger
+blocks, these failures are becoming more and more common.
+
+An important aspect of this system is that the actual challenge entropy is kept
+secret until the receipt transaction is published. This requires Hotspots to
+observe the actual data from the Challenger to correctly respond to the
+challenge and prevents poisoning attacks. However Hotspots can verify the
+challenge information against the blockchain once they see it.
+
+## Light Hotspots & GRPC
```
Comment (cokes518) [inline xxxx-validator-challenges.md:94]:
```suggestion
## Light Hotspots & gRPC
```

File: xxxx-validator-challenges.md [hunk 12]
```diff
@@ -0,0 +1,319 @@
...
+obtain better positional information as a result of the Challenge. This has been
+described in other forums as "Region" or "Multi-Hotspot" Challenges. This is out
+of scope for this HIP but we imagine future work in this direction could be
+possible.
+
+# Deployment Impact
+[deployment-impact]: #deployment-impact
+
+Overall users should see more reliable Challenge behaviour, less data use on the
+Hotspot (which makes cellular/satellite backhaul hotspots more feasible) and
```
Comment (cokes518) [inline xxxx-validator-challenges.md:293]:
```suggestion
Hotspot (which makes cellular/satellite backhaul Hotspots more feasible) and
```

File: xxxx-validator-challenges.md [hunk 13]
```diff
@@ -0,0 +1,319 @@
...
+changes to `poc_challenge_interval` as we do today in order to reduce load on
+the network.
+
+Upon handling a block, each CG member will inspect the public keys in the block,
+identify any of their own keys, and for each, initiate a new PoC. The public key
+hash will be used with the block hash to generate entropy to generate an H3
+region for the Challenge.  Entropy from a combination of the associated private
+key hash and the block hash will then be used to identify the target within the
+region to generate the Challenge itself. The region, challenge, and target will
+be persisted locally on the Challenger Validator.
```
Comment (abhay) [inline xxxx-validator-challenges.md:140]:
```suggestion
Upon handling a block, each Consensus Group member will inspect the public keys
in the block, identify any of their own keys, and for each, initiate a new PoC.
The public key hash will be used with the block hash to generate entropy to
generate an H3 region for the Challenge.  Entropy from a combination of the
associated private key hash and the block hash will then be used to identify the
target within the region to generate the Challenge itself. The region,
challenge, and target will be persisted locally on the Challenger Validator.
```

File: xxxx-validator-challenges.md [hunk 14]
```diff
@@ -0,0 +1,319 @@
...
+   synced.
+3. The Challengee then transmits the packet and sends a receipt to the
+   Challenger over the same connection from the Challenger. This is another
+   potential p2p failure.
+4. Any Hotspots observing the challenge packet then consults their local
+   blockchain ledger to see if they can resolve the challenge to a Challenger.
+   This is another potential failure point if the witness is not in sync with
+   the blockchain.
+5. If the challenge can be resolved to a Challenger then the Witness attempts to
+   dial that Challenger over p2p, another potential failure point.
```
Comment (abhay) [inline xxxx-validator-challenges.md:75]:
```suggestion
4. Any Hotspots observing the challenge packet then consults their local
   blockchain ledger to see if they can find the Challenger based on the packet
   they received and Challenge data on the blockchain. This is another potential
   failure point if the witness is not in sync with the blockchain.
5. If the Challenger is found, the Witnessing Hotspot attempts to dial them over
   p2p, another potential failure point.
```

**Pull request merged**